datafusion/common/src/scalar/logical/mod.rs

```diff
@@ -0,0 +1,400 @@
+pub use logical_duration::LogicalDuration;
+pub use logical_fixed_size_binary::LogicalFixedSizeBinary;
+pub use logical_fixed_size_list::LogicalFixedSizeList;
+pub use logical_interval::LogicalInterval;
+pub use logical_list::LogicalList;
+pub use logical_map::LogicalMap;
+pub use logical_struct::LogicalStruct;
+pub use logical_time::LogicalTime;
+pub use logical_timestamp::LogicalTimestamp;
+pub use logical_timestamp::LogicalTimestampValue;
+pub use logical_union::LogicalUnion;
+
+/// Representation of a logical scalar. Contrary to a physical [`ScalarValue`], a logical scalar
+/// is *not* coupled to a physical [`DataType`].
+///
+/// Most variants of this enum allow storing values for the variants in [`NativeType`]. Furthermore,
+/// [`LogicalScalar::Extension`] allows storing logical values for extension types.
+#[derive(Clone, Debug)]
+pub enum LogicalScalar {
+    /// A null value
+    Null,
+    /// Stores a scalar for [`NativeType::Boolean`].
+    Boolean(bool),
+    /// Stores a scalar for [`NativeType::Int8`].
+    Int8(i8),
+    /// Stores a scalar for [`NativeType::Int16`].
+    Int16(i16),
+    /// Stores a scalar for [`NativeType::Int32`].
+    Int32(i32),
+    /// Stores a scalar for [`NativeType::Int64`].
+    Int64(i64),
+    /// Stores a scalar for [`NativeType::UInt8`].
+    UInt8(u8),
+    /// Stores a scalar for [`NativeType::UInt16`].
+    UInt16(u16),
+    /// Stores a scalar for [`NativeType::UInt32`].
+    UInt32(u32),
+    /// Stores a scalar for [`NativeType::UInt64`].
+    UInt64(u64),
+    /// Stores a scalar for [`NativeType::Float16`].
+    Float16(f16),
+    /// Stores a scalar for [`NativeType::Float32`].
+    Float32(f32),
+    /// Stores a scalar for [`NativeType::Float64`].
+    Float64(f64),
+    /// Stores a scalar for [`NativeType::Timestamp`].
+    Timestamp(LogicalTimestamp),
+    /// Stores a scalar for [NativeType::Date].
+    Date(LogicalDate),
+    /// Stores a scalar for [`NativeType::Time`].
+    Time(LogicalTime),
+    /// Stores a scalar for [`NativeType::Duration`].
+    Duration(LogicalDuration),
+    /// Stores a scalar for [`NativeType::Interval`].
+    Interval(LogicalInterval),
+    /// Stores a scalar for [`NativeType::Binary`].
+    Binary(Vec<u8>),
+    /// Stores a scalar for [`NativeType::FixedSizeBinary`].
+    FixedSizeBinary(LogicalFixedSizeBinary),
+    /// Stores a scalar for [`NativeType::String`].
+    String(String),
+    /// Stores a scalar for [`NativeType::List`].
+    List(LogicalList),
+    /// Stores a scalar for [`NativeType::FixedSizeList`].
+    FixedSizeList(LogicalFixedSizeList),
+    /// Stores a scalar for [`NativeType::Struct`].
+    Struct(LogicalStruct),
+    /// Stores a scalar for [`NativeType::Union`].
+    Union(LogicalUnion),
+    /// Stores a scalar for [`NativeType::Decimal`].
+    Decimal(LogicalDecimal),
+    /// Stores a scalar for [`NativeType::Map`].
+    Map(LogicalMap),
+    // TODO logical-types: Values for ExtensionTypes
```

Comment:
I think we don't need Scalar for ExtensionTypes, they can build on top of these native scalar instead

Reply:
I think we somehow must attach the logical type of extension types to the `LogicalScalar`. The idea was to have something akin to `LogicalType::native()` that converts extension types to a `NativeType`. However, if we store the logical type in a schema we may not require such an enum variant.